### ISSUE TITLE
win32 not working with MSYS64?

### ISSUE THREAD

**Issue by GitMensch** [Bill-Gray/PDCurses#59]
Environment:

~~~
$ uname -a
MINGW64_NT-10.0 MY-MACHINE 2.5.0(0.295/5/3) 2016-03-31 18:47 x86_64 Msys

$ cc --version
cc.exe (GCC) 4.8.1
~~~

With the first try to compile (using `make -f mingwin32.mak`) the result was:

~~~
../win32/pdcscrn.c:88:5: error: unknown type name 'PCONSOLE_SCREEN_BUFFER_INFOEX'
     PCONSOLE_SCREEN_BUFFER_INFOEX lpConsoleScreenBufferInfoEx);
     ^
../win32/pdcscrn.c:90:5: error: unknown type name 'PCONSOLE_SCREEN_BUFFER_INFOEX'
     PCONSOLE_SCREEN_BUFFER_INFOEX lpConsoleScreenBufferInfoEx);
     ^
../win32/pdcscrn.c:92:1: error: unknown type name 'SetConsoleScreenBufferInfoExFn'
 static SetConsoleScreenBufferInfoExFn pSetConsoleScreenBufferInfoEx = NULL;
 ^
../win32/pdcscrn.c:93:1: error: unknown type name 'GetConsoleScreenBufferInfoExFn'
 static GetConsoleScreenBufferInfoExFn pGetConsoleScreenBufferInfoEx = NULL;
 ^
../win32/pdcscrn.c:96:1: error: unknown type name 'CONSOLE_SCREEN_BUFFER_INFOEX'
 static CONSOLE_SCREEN_BUFFER_INFOEX console_infoex;
 ^
../win32/pdcscrn.c: In function '_set_console_infoex':
../win32/pdcscrn.c:191:39: error: called object 'pSetConsoleScreenBufferInfoEx' is not a function or function pointer
     if (!pSetConsoleScreenBufferInfoEx(pdc_con_out, &console_infoex))
                                       ^
../win32/pdcscrn.c:92:39: note: declared here
 static SetConsoleScreenBufferInfoExFn pSetConsoleScreenBufferInfoEx = NULL;
                                       ^
../win32/pdcscrn.c: In function '_init_console_infoex':
../win32/pdcscrn.c:269:19: error: request for member 'cbSize' in something not a structure or union
     console_infoex.cbSize = sizeof(console_infoex);
                   ^
../win32/pdcscrn.c:271:39: error: called object 'pGetConsoleScreenBufferInfoEx' is not a function or function pointer
     if (!pGetConsoleScreenBufferInfoEx(pdc_con_out, &console_infoex))
                                       ^
../win32/pdcscrn.c:93:39: note: declared here
 static GetConsoleScreenBufferInfoExFn pGetConsoleScreenBufferInfoEx = NULL;
                                       ^
../win32/pdcscrn.c:274:19: error: request for member 'srWindow' in something not a structure or union
     console_infoex.srWindow.Right++;
                   ^
../win32/pdcscrn.c:275:19: error: request for member 'srWindow' in something not a structure or union
     console_infoex.srWindow.Bottom++;
                   ^
../win32/pdcscrn.c: In function '_get_colors':
../win32/pdcscrn.c:285:28: error: request for member 'cbSize' in something not a structure or union
         if (!console_infoex.cbSize)
                            ^
../win32/pdcscrn.c:288:42: error: request for member 'ColorTable' in something not a structure or union
             (COLORREF *)(&(console_infoex.ColorTable));
                                          ^
../win32/pdcscrn.c: In function 'PDC_scr_open':
../win32/pdcscrn.c:466:10: error: 'GetConsoleScreenBufferInfoExFn' undeclared (first use in this function)
         (GetConsoleScreenBufferInfoExFn)GetProcAddress(h_kernel,
          ^
../win32/pdcscrn.c:466:10: note: each undeclared identifier is reported only once for each function it appears in
../win32/pdcscrn.c:466:41: error: expected ';' before 'GetProcAddress'
         (GetConsoleScreenBufferInfoExFn)GetProcAddress(h_kernel,
                                         ^
../win32/pdcscrn.c:469:10: error: 'SetConsoleScreenBufferInfoExFn' undeclared (first use in this function)
         (SetConsoleScreenBufferInfoExFn)GetProcAddress(h_kernel,
          ^
../win32/pdcscrn.c:469:41: error: expected ';' before 'GetProcAddress'
         (SetConsoleScreenBufferInfoExFn)GetProcAddress(h_kernel,
                                         ^
../win32/pdcscrn.c:371:13: warning: variable 'h_kernel' set but not used [-Wunused-but-set-variable]
     HMODULE h_kernel;
             ^
../win32/pdcscrn.c: In function '_get_colors':
../win32/pdcscrn.c:296:1: warning: control reaches end of non-void function [-Wreturn-type]
 }
 ^
make: *** [pdcscrn.o] Error 1
~~~

After hacking in some option to not define `HAVE_INFOEX` the build worked but actually running any demos doesn't work.

In the MINGW64 command the result is "Redirection not supported", the result when calling from OS directly is immediately "successful" application exit (errorlevel 0).

**Comment by Bill-Gray** [comment]
We've encountered this bit of nastiness with `HAVE_INFOEX` before.  The conditions under which it is defined have been tweaked a few times.  In fact,  I see that I made a change to have `HAVE_INFOEX' #defined when compiling under MinGW back on 2016 Nov 22.  Probably a situation where it's #defined on newer versions,  but wasn't back in the days of 4.8.1.

Which leaves us with the "Redirection not supported",  which looks to be a duplicate of issue #15.  Searching for that phrase gets a few instances with various programs such as `nano` failing with it,  all when run under Cygwin.  In truth,  I don't really know what's going on here...

**Comment by mussonero** [comment]
@GitMensch  for me it worked out of the box, using
 "make -f mingwin32.mak _w64=1 WIDE=Y UTF8=Y DLL=Y ON_WINDOWS=1"

**Comment by GitMensch** [comment]
@mussonero `ON_WINDOWS=1` shouldn't be necessary as the makefile attempts to set this internal var - doesn't this work on your installation? If this is the case we should fix this.

The main problem here is not about *building* which works fine (for some versions need a hack to `HAVE_INFOEX` as mentioned above) but mainly about *running* the demos *within the MSYS64 command prompt* (the likely most version of this installed "in the wild" is the "Git bash"). Does **running** the demos inside the MSYS64 command prompt work for you?

**Comment by mussonero** [comment]
@GitMensch  My mistake, ON_WINDOWS=1 shouldn't be necessary at all as you said.
Regarding your question if the demos running inside MSYS64 terminal, the answers is no, you can only lunch them from the MSYS64 terminal successfully and they will run in new window, but not running them within the MSYS64 terminal itself.

**Comment by GitMensch** [comment]
>  you can only lunch them from the MSYS64 terminal successfully and they will run in new window, but not running them within the MSYS64 terminal itself.

@mussonero are you sure you are talking about win32 (not win32a)? If yes: how do you lunch them from MSYS64?

**Comment by mussonero** [comment]
@GitMensch  i got confused about the title "win32 not working with MSYS64?",
I was talking about Win32a, but regarding Win32 the only demo worked for me is tuidemo.exe, calling it directly from OS ,not from MSYS64 terminal.
The reset give the same result "Redirection is not supported."

**Comment by Bill-Gray** [comment]
As noted in issue #15,  this appears to be a common problem.  (Not one I've run into; I'm not using MSYS or MSYS64.)  Somehow,  I failed to note that the [message in question is in the PDCurses source](https://github.com/Bill-Gray/PDCurses/blob/master/win32/pdcscrn.c#L391),  which would be an excellent place to start looking for an answer.

We might,  for example,  stick the following test in just above the GetFileType() check:
```
if( pdc_con_in == INVALID_HANDLE_VALUE)
   perror( "Invalid input handle");
if( pdc_con_in == NULL)
   perror( "No input handle");
```
or at the very least
```
assert( pdc_con_in != INVALID_HANDLE_VALUE);
assert( pdc_con_in != NULL);
```
I get the impression this is an MSYS/MSYS64 issue,  but if it _is_ fixable on our end,  that'd be nice... it seems to have driven some people a little crazy.

**Comment by mussonero** [comment]
@Bill-Gray  interesting, after adding the above lines before GetFileType() and lunching the demos from MSYS64 terminal, i am still getting the same previews massage "Redirection is not supported." nothing else, but lunching the demos directly from OS gives "Assertion fails Expression pdc_con_in != NULL"

**Comment by Bill-Gray** [comment]
@mussonero That _is_ interesting.... [Microsoft's documentation for GetStdHandle](https://docs.microsoft.com/en-us/windows/console/getstdhandle) says:
```
If an application does not have associated standard handles, such as a service running
on an interactive desktop, and has not redirected them, the return value is NULL.
```
So in _both_ cases,  redirection isn't supported.  Launch directly from the OS,  and that's indicated as described above,  with a NULL handle.  Go from the MSYS64 terminal,  and you get a valid handle,  but can't redirect it.

I don't know much about the Windows console,  and can't say why this should happen in some configurations but not others or how we'd fix/work our way around the problem.

**Comment by GitMensch** [comment]
What happens if you add a manual redirection?

We may be able to aquire the handles and do so only if we get NULL in the first check.

**Comment by mussonero** [comment]
@GitMensch @Bill-Gray , another interesting fact, by removing "-mwindows" flag from mingwin32.mak under Win32, make the demos run directly from OS with no issues of "Redirection is not supported." .

**Comment by Bill-Gray** [comment]
@mussonero - Thank you;  I think you've fixed both this issue and #67.

Furthermore,  I just tried running the Win32 console demos on an old WinME box (using MinGW straight from command.com,  no Cygwin involvement),  and found that I had to remove the -mwindows switch there,  too.

The gcc man page says
```
-mwindows
This option is available for Cygwin and MinGW targets.  It specifies that a GUI
application is to be generated by instructing the linker to set the PE header
subsystem type appropriately.
```
The `-mwindows` option was added to the Win32a (GUI) MinGW makefile in mid-2013,  and then made its way into the Win32 (console) makefile a few years later,  when I revised things to make those files as similar as reasonably possible.  I am reasonably sure that `-mwindows` belongs in Win32a,  but not in the Win32 console makefile,  and have just pushed a fix that removes `-mwindows` from the latter.

For that matter,  I'm not entirely sure if `-mwindows` needs to be present in the Win32a (GUI) version of mingwin32.mak.  I tried removing it,  both cross-compiling on Linux and running in Wine,  and on the old WinME box,  and it didn't seem to make a difference.  I'm not going to rush to remove it,  since it appears to do no harm.  But I'm wondering why it's there.

**Comment by GitMensch** [comment]
@Bill-Gray The difference of the subsystem in the PE header is normally important for the question if you see a (likely empty) command prompt if you double click on the executable file or not. Normally nearly all applications that use GUI windows should set it but it must be tested if sysout/syserr still work as expected (which is itself mostly up to definition - many applications just "loose" this function [you will never see anything that is sent to sysout/syserr] as long as you don't run the application from command line [where you will see it again and can redirect it if you want to], some loose it completely and some just create a file for each [I think most/all SDL apps do this]).

But it is definitely not to set on pure console applications. I try to find the time to recheck that everything win32 works with some Windows machines in the upcoming week and close both tickets afterwards.
